// @anonaddy/omni-worker/dist/cjs/node/builder/helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildWorkerCode = buildWorkerCode;
const webpack_node_externals_1 = __importDefault(require("webpack-node-externals"));
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const webpack_1 = require("webpack");
const terser_webpack_plugin_1 = __importDefault(require("terser-webpack-plugin"));
const BUILD_PATH = path_1.default.join(__dirname, '.out');
const BUILD_FILENAME = 'bundle.js';
const BUILD_FILE_PATH = path_1.default.join(BUILD_PATH, BUILD_FILENAME);
function buildWorkerCode(entryFile) {
    if (!fs_1.default.existsSync(BUILD_PATH)) {
        fs_1.default.mkdirSync(BUILD_PATH, { recursive: true });
    }
    const config = getWebpackConfig(entryFile);
    const compiler = (0, webpack_1.webpack)(config);
    return new Promise((resolve, reject) => {
        compiler.run((err, stats) => {
            if (err) {
                reject(err);
                return;
            }
            /*
            console.log(stats?.toString({
              chunks: false,
              colors: true
            }));
            */
            const bundledCode = fs_1.default
                .readFileSync(BUILD_FILE_PATH, 'utf8');
            // clean up the build folder
            fs_1.default.rm(BUILD_PATH, { recursive: true, force: false }, (err) => {
                if (err) {
                    reject(err);
                }
            });
            resolve(bundledCode);
        });
    });
}
const getWebpackConfig = (entry) => ({
    entry,
    output: {
        filename: BUILD_FILENAME,
        path: BUILD_PATH,
        libraryTarget: 'commonjs2',
        module: false,
        iife: false,
        umdNamedDefine: true,
    },
    resolve: {
        extensions: ['.ts', '.js'],
        fallback: {
            assert: false,
            buffer: false,
            console: false,
            constants: false,
            crypto: false,
            domain: false,
            events: false,
            fs: false,
            http: false,
            https: false,
            module: false,
            net: false,
            os: false,
            path: false,
            process: false,
            punycode: false,
            querystring: false,
            readline: false,
            stream: false,
            string_decoder: false,
            sys: false,
            timers: false,
            tls: false,
            tty: false,
            url: false,
            util: false,
            vm: false,
            worker_threads: false,
            zlib: false,
        },
    },
    resolveLoader: {
        modules: [
            path_1.default.resolve(process.cwd(), 'node_modules')
        ]
    },
    mode: 'production',
    optimization: {
        minimize: true,
        minimizer: [
            new terser_webpack_plugin_1.default({
                terserOptions: {
                    compress: {
                        drop_console: true,
                        drop_debugger: true
                    },
                }
            })
        ]
    },
    target: 'node',
    module: {
        rules: [
            {
                test: /\.ts$/,
                use: [
                    {
                        loader: resolveNodeModules('babel-loader'),
                        options: {
                            presets: [
                                resolveNodeModules('@babel/preset-env'),
                                resolveNodeModules('@babel/preset-typescript')
                            ],
                            plugins: [
                                resolveNodeModules('@babel/plugin-transform-modules-commonjs'),
                                [resolveNodeModules('@babel/plugin-proposal-decorators'), { "legacy": true }],
                            ]
                        },
                    }
                ],
                exclude: /node_modules/
            }
        ],
    },
    devtool: 'source-map',
    stats: 'verbose',
    externals: [(0, webpack_node_externals_1.default)({
            importType: 'commonjs',
        })],
    cache: false
});
const resolveNodeModules = (packageName) => require.resolve(packageName, {
    paths: [
        path_1.default.resolve(__dirname, 'node_modules')
    ]
});
