// @anonaddy/omni-worker/dist/mjs/node/index.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __setFunctionName = (this && this.__setFunctionName) || function (f, name, prefix) {
    if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
    return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
};
import { parentPort } from 'worker_threads';
import Comlink from 'comlink';
import nodeEndpoint from 'comlink/dist/umd/node-adapter';
import { buildNodeApiAndWorkerFromCode, genWorkerCodeFromFile } from './builder';
import { staticImplements } from '../types/helpers';
/**
 * OmniWorker for NodeJS
 */
let NodeOmniWorker = (() => {
    let _classDecorators = [staticImplements(), staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var NodeOmniWorker = _classThis = class {
        constructor(code, worker, api) {
            this.isInitialized = () => (this._api !== undefined);
            this.use = () => {
                const isInitialized = this.isInitialized();
                if (isInitialized) {
                    return this._api;
                }
                else {
                    throw Error('worker is not yet initialized. make sure to call the build() function, first');
                }
            };
            this.clone = (numOfTimes) => {
                const workers = [];
                const code = this._code;
                for (let i = 0; i <= numOfTimes; i++) {
                    const { worker, api } = buildNodeApiAndWorkerFromCode(code);
                    workers.push(new NodeOmniWorker(code, worker, api));
                }
                return workers;
            };
            this.destroy = async () => {
                await this._worker.terminate();
            };
            this._code = code;
            this._worker = worker;
            this._api = api;
            return this;
        }
        /**
         * Handles the building of the necessary resources for an OmniWorker to function.
         * This usually means, building from the consumer's code, creating the comlink
         * interface between the worker and the main thread as well as the worker itself.
         *
         * @param path Relative path FROM YOUR PROJECT's ROOT to the file to be the worker
         * @returns An OmniWorker
         */
        static async build(path) {
            const code = await genWorkerCodeFromFile(path);
            const { worker, api } = buildNodeApiAndWorkerFromCode(code);
            return new NodeOmniWorker(code, worker, api);
        }
    };
    __setFunctionName(_classThis, "NodeOmniWorker");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        NodeOmniWorker = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
    })();
    /**
     * Expose the functions inside the worker to the rest of the application.
     * After having exposed the functions, the build step can be initiated.
     * @param exposable A class, object with functions or a single function to be
     * exposed to the main thread.
     */
    _classThis.expose = (exposable) => {
        if (parentPort) {
            Comlink.expose(exposable, nodeEndpoint(parentPort));
        }
    };
    (() => {
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return NodeOmniWorker = _classThis;
})();
export { NodeOmniWorker };
