// @anonaddy/omni-worker/dist/mjs/node/pool.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __setFunctionName = (this && this.__setFunctionName) || function (f, name, prefix) {
    if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
    return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
};
import { NodeOmniWorker } from ".";
import { staticImplements } from "../types/helpers";
let NodeOmniWorkerPool = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var NodeOmniWorkerPool = _classThis = class {
        constructor(from, options) {
            this._pool = [];
            this._count = 0;
            this._lastUseIdx = -1;
            this.isInitialized = () => this._count > 0;
            this.getNumOfWorkers = () => this._count;
            this.use = () => {
                if (!this.isInitialized()) {
                    return this._from.use();
                }
                else {
                    const lastIdx = this._lastUseIdx;
                    // first run or maxed out
                    if (lastIdx === this._count - 1 ||
                        lastIdx === -1) {
                        this._lastUseIdx = 0;
                        return this._pool[0].use();
                        // any other run
                    }
                    else {
                        this._lastUseIdx = lastIdx + 1;
                        return this._pool[lastIdx + 1].use();
                    }
                }
            };
            this.destroy = async () => {
                const fns = [];
                this._pool.forEach(w => {
                    const fn = async () => {
                        await w.destroy();
                    };
                    fns.push(fn);
                });
                await Promise.all(fns.map(fn => fn()));
            };
            this._from = from;
            this._options = options;
            this.applyOptions();
        }
        /**
         * First builds the OmniWorkers and then launches a new NodeOmniWorker pool
         * @param from The (relative) file path from your project's root to the worker .ts file
         * @param options (Optional) Options object for launching the pool
         */
        static async buildAndLaunch(from, options = {
            numOfWorkers: 1
        }) {
            const worker = await NodeOmniWorker.build(from);
            const pool = NodeOmniWorkerPool.launch(worker, options);
            return pool;
        }
        /**
         * Launches a new NodeOmniWorker pool according to the options provided
         * @param worker An already built NodeOmniWorker
         * @returns A newly created pool
         */
        static launch(worker, options = {
            numOfWorkers: 1
        }) {
            const pool = new NodeOmniWorkerPool(worker, options);
            return pool;
        }
        /**
         * Applies the options provided
         */
        applyOptions() {
            const { numOfWorkers } = this._options;
            let count = 0;
            if (numOfWorkers === undefined) {
                count = 1;
            }
            else {
                count = numOfWorkers;
            }
            if (count < 1) {
                throw Error(`number of workers must be at least 1`);
            }
            this._count = count;
            for (let i = 0; i <= count; i++) {
                this._pool = this._from.clone(count);
            }
        }
    };
    __setFunctionName(_classThis, "NodeOmniWorkerPool");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        NodeOmniWorkerPool = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return NodeOmniWorkerPool = _classThis;
})();
export { NodeOmniWorkerPool };
